### ISSUE TITLE
Update Font Awesome

### ISSUE THREAD

**Issue by SoftCreatR** [WoltLab/WCF#2496]
Font Awesome 5 has been finally released. Please consider updating the whole FA implementation in WSC 3.1.

https://fontawesome.com

It contains 218 new icons (in the free version, compared to FA 4.7) and it *should* be easier to implement and maintain as it was before.

There's also a way for a "simple" update, using a shim. See https://fontawesome.com/how-to-use/upgrading-from-4 for more informations.

**Comment by mutec** [comment]
Using CSS/SCSS is more complicated because it's divided into `fas` (filled, non-o), `far` (not filled, o-types), `fab` (brands) and `fal` (light, pro only). Some of the regular icons are pro only too, but available within solid for free (e.g. the bell).
This might cause incompatibilities and angry devs.

Using JS and SVG (the simple update thing you mentioned) forces the browser to render a lot of SVGs (not sure what that means for the performance). Another point against JS is, that the code has to be finalized at the moment the page get's rendered and usage within AJAX-Responses, other JS could cause problems or more work.

I don't see any "easy" way to implement this without a fallback to 4.7 by loading 4 font-faces (4.7 by default, and the three new ones if their new classes are used).
My first idea was [this commit](https://github.com/mutec/WCF/commit/eae2e262f686e23aa4cd217ec632eee0e77f6041), but I don't like that way of implementation.

**Comment by dtdesign** [comment]
I've spent quite some time looking into the technical implications of FontAwesome 5 and I came to the conclusion to *not migrate* to FontAwesome 5 at this time.

What has changed from FontAwesome 4 to FontAwesome 5:
 1. There is no longer a single font, but instead there are 3 fonts:
   1. `regular` and `solid` both offer somewhat the same icons (with the same char code!), but using a different appearance.
   2. `brands` is the collection of brand-specific icons.
 2. Icons that feature both a solid-color and non-solid variant now share the same char code, but reside in either the `regular` or the `solid` font. For instance, the bookmark icon in FA4 was `\f02e` (solid) and `\f097` (non-solid), now both are `\f02e` and `\f097` is no longer being used.
 3. In FA4 all icons went by the CSS classes `fa` and the `fa-<icon>`, but in FA5 we have these ones:
  1. `fab` => `Font Awesome Brands`
  2. `fas` _or_ `fa` => `Font Awesome Solid`
  3. `far` => `Font Awesome Regular`
  4. `fal` => `Font Awesome Light` (FA5 Pro only)

To make things even more complicated, the `brands` font uses the font-family `FontAwesome 5 Brands`, but both the `solid` and `regular` font goes by the name `FontAwesome 5 Free`. Yep, that is right, it is the same exact font name, but (!) they differ by the font-weight. Solid icons go with the value `900` and regular is recognized by the value `normal` (which equals `400`).

At the end of the day, I would need to manipulate the CSS before it is finally saved in order to rewrite the FA4-class-usage by adjusting the `content`-, `font-family`- and `font-weight`-properties for each occurrence, while carefully trying not to break FA5 usages.

It gets even worse because we can't change the CSS class names, such as the `.icon`/`.fa` class, in order to preserve backwards compatibility. Otherwise this would create a real mess in the CSS where the result is highly unpredictable and prone to errors due to the fact that the `fa` class is still used. It would have been a lot easier if they ditched the `fa` class, so I could perfectly scope the changes to legacy class usages only.

Long story short: Supporting FontAwesome 5 is not an option at this time, especially for backwards compatibility reasons. Allthough I'm confident that we could find a solution in the long-run, there is no easy and reliable fix that I can pull out at the mere end of the beta phase. We're looking forward to push the product towards the stable RC phase, not to break everything out of nowhere.

PS: Their JavaScript shim is… not sufficient. It may work nicely with static HTML pages, but there is quite often a deeper JavaScript-based logic attached to the CSS class names and the DOM. Even worse, not all of these shortcomings are blatantly sticking out, most of them are taking place in the background, where you won't notice it, unless you hit a specific function that relies on it.

**Comment by dtdesign** [comment]
One more thing: This isn't just about adding support for the legacy classes (such as `.fa-bookmark-o`), I also have to take into account that the icons are used outside of their classes. This includes constructs such as `content: $fa-var-<icon>;` in combination with `font-family: FontAwesome`, which appear both in our and in 3rd party products. These need to be handled when processing the CSS, because I would need to alter the output code before it is written to the stylesheet. Making these manual adjustments in a static SCSS file really isn't much of a big deal, annoying but not overly complicated.
